[cookie-parser] Add secret to express's request

mckramer opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/expressjs/cookie-parser/blob/7998537f808b5502f1738b844cdf2c56dc9459b2/index.js#L51

This Express middleware conditionally sets the secret field on Express's Request.  It can then be used by other libraries, such as csurf here: https://github.com/expressjs/csurf/blob/master/index.js#L272.

BendingBender commented on types/cookie-parser/cookie-parser-tests.ts:
```diff
@@ -4,6 +4,10 @@ import cookieParser = require('cookie-parser');
 const app = express();
 app.use(cookieParser('optional secret string'));
 
+app.use((req: express.Request) => {
+    const secret: string | undefined = req.secret;
+});
```
This is better tested like this:
```suggestion
const req = null as any as express.Request;
req.secret; // $ExpectType string | undefined
```

BendingBender commented on types/cookie-parser/index.d.ts:
```diff
@@ -7,6 +7,20 @@
+             * Optionally set by cookie-parser if secret(s) are provided.  Can be used by other middleware.
+             * [Declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) can be used to add your own properties.
+             */
+            secret?: string;
```
There are more properties added to the Request object:

- `cookies: Record<string, string>`
- `signedCookies: Record<string, string | Record<string,unknown> | false>`

mckramer replied:
They are technically part of the core express request object here, but as "any" type.  I wasn't sure if desire would be to override the types here, but would happy to add and try it out.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express-serve-static-core/index.d.ts#L622
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express-serve-static-core/index.d.ts#L632

Thoughts?

BendingBender replied:
Yeah, go for it. Ping me if you encounter any problems.

mckramer replied:
Thanks for help, I just pushed the defs for cookies/signedCookies, but test is failing locally.  Two issues:

1. It seems to inherit the jsdoc, but not the more specific types.  Still returning as `any` in the test (and VS Code).
1. Are we sure the types are specific enough, given that JSON cookies are a possibility as well?  e.g. `Record<string, string>` seems like would be too restrictive for `cookies`.

Let me know if you have any thoughts on how to override the types (and ensure they are correct), or if something should that be discussed in `express-service-static-core`?

BendingBender replied:
You're absolutely right, `cookies` must have the type `Record<string, unknown>`. Basically, `signedCookies` should have the same type. But this will break lots of users. Now that I think about it, just remove the two, this is too big and too breaking.

mckramer commented:
> @BendingBender Thank you for reviewing this PR! The author has pushed new commits since your last review. Could you take another look and submit a fresh review?

Removed cookies/signedCookies per conversation and squashed + rebased change to latest master.

mckramer commented:
Ready to merge
